### ISSUE TITLE
[Feature Request] provide abstractions to allow alternative debugger hookups (DAP) to use termdebug UX

### ISSUE THREAD

**Issue by jinyius** [vim/vim#11605]
**Is your feature request about something that is currently impossible or hard to do? Please describe the problem.**
termdebug supports GDB/MI only.  [DAP](https://microsoft.github.io/debug-adapter-protocol/) exists.

**Describe the solution you'd like**
It would be great if the UX around termdebug could be reused for an ambitious plugin writer to support working with DAP debuggers.

**Describe alternatives you've considered**
vimspector and neovim's nvim-dap

**Additional context**
nope

**Comment by puremourning** [comment]
> Describe alternatives you've considered
vimspector and neovim's nvim-dap

Do they not count as "ambitious" enough ?

**Comment by razamatan** [comment]
vimL only, no lua or python3 requirement
________________________________

**Comment by brammool** [comment]
From the front page DAP looks like a good idea to support.   Perhaps start with one debugger that supports it and try make it work with TermDebug?  Then plan to expand to support more.  I have no idea how much work this would be.  I can see a protocol specification, but it doesn't say how the communication is implemented.  Using JSON?
It does appear to be very MS-Windows specific, character units are UTF-16 (worst choice ever).

**Comment by puremourning** [comment]
>  I can see a protocol specification, but it doesn't say how the communication is implemented. Using JSON?

It uses something like JSON-RPC over stdio. https://microsoft.github.io/debug-adapter-protocol/overview#base-protocol
basically a length header followed by a json payload.

> I have no idea how much work this would be.

It's not a small amount of work, in and of itself, but the ecosystem of debug adapters and foibles is a lot of ongoing maintenance. I have been working on vimspector https://github.com/puremourning/vimspector/ for 5 years now and it implements about, say 40% of the spec. Implementing [disassembly](https://github.com/puremourning/vimspector/#disassembly) was weeks of work (elapsed, spare time) on its own.

it's also worth noting there are a number of other DAP plugins for vim that are very good.

**Comment by razamatan** [comment]
yeah, i took a good look at vimspector (listed in original post as alternatives considered.. tyvm for it, btw), but sometimes i don't have access to a +python3 compiled vim readily.  that got me looking around and land on termdebug as a built-in thing.

it would be great if the abstractions existed since vimspector could reuse the same windowing approach and command bindings.  i'm not deep enough down the rabbit hole of how vimspector is implemented to appreciate how difficult it would be to make it work at the window/UX layer to interop with the termdebug abstractions (if such things existed).  but that's the FR.  not suggesting we put a competing DAP client out there, but allow existing DAP clients to use more of the built-in debug experience.
________________________________

**Comment by kohnish** [comment]
GDB has DAP support now. I'm curious if we can try to give an optional support from termdebug already.

**Comment by ubaldot** [comment]
I also support this. Now that termdebug is in Vim9 it should be easier to update. As a first step I would try to separate the frontend to the backend (which now only include gdb). In that way the frontend would be always the same but the backend should be easy to switch i.e. it should be easier to add other debugger support like debugpy, lldb, etc. or for a possible DAP implementation entirely in Vim9.

**Comment by ubaldot** [comment]
@jinyius Given that DAP is the main wish, could you edit the title such that such a desire is more explicit? In this way there would be higher chances that someone could pick it up. Or one could also consider a porting of nvim-dap to Vim9 if does not want to change Termdebug.
